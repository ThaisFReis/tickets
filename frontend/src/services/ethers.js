import { ethers } from 'ethers';
import contractAbi from '../contract-abi.json';
import contractAddress from '../contract-address.json';
import { eventLocations } from '../locations-data';
import { eventsMetadata } from '../events-metadata';

let provider;
if (window.ethereum) {
  provider = new ethers.BrowserProvider(window.ethereum);
} else {
  console.error("Please install MetaMask!");
  provider = new ethers.JsonRpcProvider('http://localhost:8545');
}

const getSigner = async () => {
  if (!window.ethereum) {
    throw new Error("MetaMask is not installed.");
  }
  const signer = await provider.getSigner();
  return signer;
};

const getContract = async (withSigner = false) => {
  try {
    if (withSigner) {
      const signer = await getSigner();
      return new ethers.Contract(contractAddress.address, contractAbi, signer);
    }
    const localProvider = new ethers.JsonRpcProvider('http://localhost:8545');
    return new ethers.Contract(contractAddress.address, contractAbi, localProvider);
  } catch (error) {
    console.error("Error getting contract:", error);
    throw error;
  }
};

const createEvent = async (name, date, tierNames, tierPrices, tierQuantities) => {
  try {
    const contract = await getContract(true);
    const pricesInWei = tierPrices.map(price => ethers.parseEther(price));
    const transaction = await contract.createEvent(
      name,
      date,
      tierNames,
      pricesInWei,
      tierQuantities
    );
    await transaction.wait();
  } catch (error) {
    console.error("Error creating event:", error);
    throw error;
  }
};

const purchaseTickets = async (eventId, tier, quantity, totalPrice) => {
  try {
    const contract = await getContract(true);
    
    if (!tier || !tier.tierId) {
      throw new Error(`Invalid tier object provided for event ${eventId}`);
    }
    const tierId = tier.tierId;

    const transaction = await contract.purchaseTickets(eventId, tierId, quantity, { value: totalPrice });
    const receipt = await transaction.wait();
    return receipt;
  } catch (error) {
    console.error("Error buying ticket:", error);
    throw error;
  }
};

const fetchAllEvents = async () => {
  try {
    const contract = await getContract();
    const eventCount = await contract.nextEventId();
    const events = [];
    for (let i = 1; i < eventCount; i++) {
      const eventDetails = await contract.getEventDetails(i);
      const tiers = await getEventTiers(i, eventDetails); // Pass details to avoid re-fetching
      events.push({
        eventId: BigInt(i),
        name: eventDetails[0],
        date: eventDetails[1],
        organizer: eventDetails[2],
        tiers: tiers,
      });
    }
    return events;
  } catch (error) {
    console.error("Error fetching all events:", error);
    throw error;
  }
};

const getEventDetails = async (eventId) => {
  try {
    const contract = await getContract();
    return await contract.getEventDetails(eventId);
  } catch (error) {
    console.error("Error getting event details:", error);
    throw error;
  }
};

const getEventTiers = async (eventId, eventDetails) => {
  try {
    const contract = await getContract();
    // If eventDetails are not provided, fetch them. This makes the function more robust.
    if (!eventDetails) {
      eventDetails = await contract.getEventDetails(eventId);
    }
    const eventName = eventDetails[0];
    const eventMeta = eventsMetadata.find(e => e.name === eventName);
    const locationMeta = eventLocations.find(l => l.name === eventMeta?.location);

    const tiers = [];
    let i = 1;
    while (true) {
      try {
        const tier = await contract.getTicketTier(eventId, i);
        const tierName = tier.name;
        const tierType = locationMeta?.seating[tierName]?.type || 'standing';

        tiers.push({
          tierId: Number(tier.typeId),
          name: tierName,
          price: tier.price.toString(),
          totalQuantity: Number(tier.quantity),
          sold: Number(tier.sold),
          type: tierType,
        });
        i++;
      } catch {
        break; // Assumes error means no more tiers
      }
    }
    return tiers;
  } catch (error) {
    console.error("Error fetching event tiers:", error);
    throw error;
  }
};

const getTicketsOfOwner = async (owner) => {
  try {
    const contract = await getContract();
    return await contract.getTicketsOfOwner(owner);
  } catch (error) {
    console.error("Error getting tickets of owner:", error);
    throw error;
  }
};

const isSeatSold = async (eventId, seatId) => {
  try {
    const contract = await getContract();
    return await contract.isSeatSold(eventId, seatId);
  } catch (error) {
    console.error("Error checking if seat is sold:", error);
    throw error;
  }
};

const getOwnerOfTicket = async (tokenId) => {
  try {
    const contract = await getContract();
    return await contract.ownerOf(tokenId);
  } catch (error) {
    console.error("Error getting owner of ticket:", error);
    throw error;
  }
};

const getTokenURI = async (tokenId) => {
  try {
    const contract = await getContract();
    return await contract.tokenURI(tokenId);
  } catch (error) {
    console.error("Error getting token URI:", error);
    throw error;
  }
};

const connectWallet = async () => {
  if (!window.ethereum) {
    throw new Error("MetaMask is not installed. Please install it to use this app.");
  }
  try {
    const accounts = await provider.send("eth_requestAccounts", []);
    const address = accounts[0];
    const signer = await provider.getSigner();
    return { signer, address };
  } catch (error) {
    console.error("Failed to connect wallet:", error);
    throw new Error("Wallet connection was rejected.");
  }
};

const fetchUserTickets = async (userAddress) => {
  try {
    const contract = await getContract();
    const ticketIds = await contract.getTicketsOfOwner(userAddress);
    
    const tickets = await Promise.all(
      ticketIds.map(async (tokenId) => {
        const details = await contract.getTicketDetails(tokenId);
        return {
          tokenId: tokenId.toString(),
          eventId: details.eventId.toString(),
          tierName: details.tierName,
          seatId: details.seatId,
        };
      })
    );
    return tickets;
  } catch (error) {
    console.error("Error fetching user tickets:", error);
    return [];
  }
};

const fetchSoldSeatsForTier = async (eventId, tierName) => {
  try {
    const contract = await getContract();
    const eventDetails = await contract.getEventDetails(eventId);
    const eventName = eventDetails[0];
    
    const eventMeta = eventsMetadata.find(e => e.name === eventName);
    if (!eventMeta) return [];

    const locationMeta = eventLocations.find(l => l.name === eventMeta.location);
    if (!locationMeta || !locationMeta.seating[tierName] || !locationMeta.seating[tierName].layout) {
      return []; // Not a seated tier or no layout defined
    }

    const layout = locationMeta.seating[tierName].layout;
    
    // Find the tierId that corresponds to the tierName
    const tiers = await getEventTiers(eventId, eventDetails);
    const tier = tiers.find(t => t.name === tierName);
    if (!tier) return [];
    const tierId = tier.tierId;

    const soldSeats = [];
    for (let i = 0; i < layout.rows; i++) {
      for (let j = 0; j < layout.seatsPerRow; j++) {
        // This seatId construction must match the one in SeatSelectionMap.jsx
        const seatIdString = `${tierName}-${i}-${j}`; 
        // The contract needs a numerical seatId. We can derive it from the loop indices.
        const seatIndex = i * layout.seatsPerRow + j; 
        
        const owner = await contract.getSeatOwner(eventId, tierId, seatIndex);
        if (owner !== '0x0000000000000000000000000000000000000000') {
          soldSeats.push(seatIdString);
        }
      }
    }
    return soldSeats;
  } catch (error) {
    console.error("Error fetching sold seats:", error);
    return [];
  }
};

export {
  provider,
  getSigner,
  getContract,
  connectWallet,
  createEvent,
  purchaseTickets,
  fetchAllEvents,
  fetchUserTickets,
  fetchSoldSeatsForTier,
  getEventDetails,
  getEventTiers,
  getTicketsOfOwner,
  isSeatSold,
  getOwnerOfTicket,
  getTokenURI,
};
