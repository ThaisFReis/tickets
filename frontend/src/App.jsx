import React, { useState, useEffect } from 'react';
import Header from './components/Header';
import HomePage from './pages/HomePage';
import EventDetailPage from './pages/EventDetailPage';
import ProfilePage from './pages/ProfilePage';
import TransactionModal from './components/TransactionModal';
import Footer from './components/Footer';
import { fetchAllEvents, connectWallet, purchaseTickets, fetchUserTickets } from './services/ethers';
import { eventsMetadata } from './events-metadata';
import './index.css';
import EventCardSkeleton from './components/EventCardSkeleton';

function App() {
    const [currentPage, setCurrentPage] = useState('home');
    const [selectedEvent, setSelectedEvent] = useState(null);
    const [walletConnected, setWalletConnected] = useState(false);
    const [userAddress, setUserAddress] = useState('');
    const [ownedTickets, setOwnedTickets] = useState([]);
    const [events, setEvents] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);
    
    const [modalState, setModalState] = useState('closed');
    const [purchaseDetails, setPurchaseDetails] = useState(null);

    useEffect(() => {
        const loadEvents = async () => {
            try {
                setLoading(true);
                const fetchedEvents = await fetchAllEvents();
                
                const eventsWithDetails = fetchedEvents.map(event => {
                    const metadata = eventsMetadata.find(m => m.name === event.name) || {};
                    const dateObj = new Date(Number(event.date) * 1000);

                    const year = dateObj.getFullYear();
                    const month = String(dateObj.getMonth() + 1).padStart(2, '0');
                    const day = String(dateObj.getDate()).padStart(2, '0');
                    const localDateString = `${year}-${month}-${day}`;

                    return {
                        ...event,
                        id: Number(event.eventId),
                        date: dateObj.toISOString(),
                        localDate: localDateString,
                        ...metadata,
                    };
                });

                setEvents(eventsWithDetails);
                setError(null);
            } catch (err) {
                console.error("Error fetching events from contract:", err);
                setError("Failed to load events. Please ensure you're connected to the correct network.");
            } finally {
                setLoading(false);
            }
        };
        loadEvents();
    }, []);

    useEffect(() => {
        const handleAccountsChanged = (accounts) => {
            if (accounts.length > 0) {
                const address = accounts[0];
                setUserAddress(address);
                setWalletConnected(true);
            } else {
                setWalletConnected(false);
                setUserAddress('');
                setOwnedTickets([]);
            }
        };

        const handleChainChanged = () => {
            window.location.reload();
        };

        if (window.ethereum) {
            window.ethereum.on('accountsChanged', handleAccountsChanged);
            window.ethereum.on('chainChanged', handleChainChanged);

            const checkConnection = async () => {
                try {
                    const accounts = await window.ethereum.request({ method: 'eth_accounts' });
                    if (accounts.length > 0) {
                        handleAccountsChanged(accounts);
                    }
                } catch (error) {
                    console.error("Error checking for wallet connection:", error);
                }
            };
            checkConnection();
        }

        return () => {
            if (window.ethereum) {
                window.ethereum.removeListener('accountsChanged', handleAccountsChanged);
                window.ethereum.removeListener('chainChanged', handleChainChanged);
            }
        };
    }, []);

    useEffect(() => {
        const loadUserTickets = async () => {
            if (userAddress && events.length > 0) {
                try {
                    const tickets = await fetchUserTickets(userAddress);
                    const ticketsWithMetadata = tickets.map(ticket => {
                        const event = events.find(e => e.id === Number(ticket.eventId));
                        if (!event) return null;
                        return { 
                            ...ticket, 
                            eventName: event.name, 
                            eventDate: event.date, 
                            eventImage: event.image, 
                            venue: event.location 
                        };
                    }).filter(Boolean);
                    setOwnedTickets(ticketsWithMetadata);
                } catch (error) {
                    console.error("Failed to fetch user tickets:", error);
                }
            } else {
                setOwnedTickets([]);
            }
        };
        loadUserTickets();
    }, [userAddress, events]);

    const handleConnectWallet = async () => {
        if (walletConnected) {
            setWalletConnected(false);
            setUserAddress('');
            setOwnedTickets([]);
        } else {
            try {
                const { address } = await connectWallet();
                setUserAddress(address);
                setWalletConnected(true);
            } catch (error) {
                console.error("Failed to connect wallet:", error);
                alert("Failed to connect wallet. Please make sure MetaMask is installed and unlocked.");
            }
        }
    };
    
    const handleSelectEvent = (event) => {
        setSelectedEvent(event);
        setCurrentPage('event');
        window.scrollTo(0, 0);
    };
    
    const handleNavigate = (page) => {
        if (page === 'profile' && !walletConnected) {
            alert('Por favor, conecte a sua carteira para ver os seus ingressos.');
            return;
        }
        setCurrentPage(page);
        setSelectedEvent(null);
        window.scrollTo(0, 0);
    };

    const handleNavigateBack = () => {
        setCurrentPage('home');
        setSelectedEvent(null);
    }

    const handlePurchase = (tier, quantity, totalPrice) => {
        setPurchaseDetails({ 
            event: selectedEvent, 
            tier,
            quantity,
            totalPrice
        });
        setModalState('confirm');
    };

    const handleConfirmPurchase = async () => {
        if (!purchaseDetails) return;
        setModalState('processing');
        try {
            const { event, tier, quantity, totalPrice } = purchaseDetails;
            
            await purchaseTickets(event.id, tier, quantity, totalPrice);

            setModalState('success');

            const updatedTickets = await fetchUserTickets(userAddress);
             const ticketsWithMetadata = updatedTickets.map(ticket => {
                const event = events.find(e => e.id === Number(ticket.eventId));
                if (!event) return null;
                return { 
                    ...ticket, 
                    eventName: event.name, 
                    eventDate: event.date, 
                    eventImage: event.image, 
                    venue: event.location 
                };
            }).filter(Boolean);
            setOwnedTickets(ticketsWithMetadata);
            
            setTimeout(() => {
                setModalState('closed');
                setCurrentPage('profile');
                window.scrollTo(0, 0);
            }, 3000);
        } catch (error) {
            console.error("Purchase failed:", error);
            setModalState('error');
        }
    };
    
    const renderPage = () => {
        switch (currentPage) {
            case 'event':
                return <EventDetailPage 
                            event={selectedEvent} 
                            onPurchase={handlePurchase} 
                            walletConnected={walletConnected}
                            onNavigateBack={handleNavigateBack}
                        />;
            case 'profile':
                return <ProfilePage tickets={ownedTickets} userAddress={userAddress} />;
            case 'home':
            default:
                return <HomePage events={events} onSelectEvent={handleSelectEvent} isLoading={loading} />;
        }
    };

    return (
        <div className="bg-background min-h-screen flex flex-col">
            <div className="gradient-bg">
                <div className="shape1"></div>
                <div className="shape2"></div>
            </div>
            <Header
                walletConnected={walletConnected}
                userAddress={userAddress}
                onConnectWallet={handleConnectWallet}
                onNavigate={handleNavigate}
            />
            <main className="flex-grow">
                {renderPage()}
            </main>
            <TransactionModal 
                modalState={modalState}
                setModalState={setModalState}
                purchaseDetails={purchaseDetails}
                onConfirm={handleConfirmPurchase}
            />
            <Footer />
        </div>
    );
}

export default App;
