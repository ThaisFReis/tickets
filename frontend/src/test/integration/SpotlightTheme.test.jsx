import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import App from '../../App';
import * as EthersService from '../../services/ethers';
import { vi } from 'vitest';

// Mock the ethers service to prevent contract calls during snapshot testing
vi.mock('../../services/ethers');

describe('Spotlight Theme Snapshot Test', () => {
  it('renders the application with the new Spotlight theme', async () => {
    // Mock the event fetching to provide consistent data
    const mockEvents = [
      { eventId: 1n, name: 'Rock Fest 2025', date: BigInt(Math.floor(Date.now() / 1000) + 86400) },
    ];
    EthersService.fetchAllEvents.mockResolvedValue(mockEvents);

    const { asFragment } = render(<App />);
    
    // This will create a snapshot of the App component's DOM structure and styles
    // The first run will create the snapshot file. Subsequent runs will compare against it.
    expect(asFragment()).toMatchSnapshot();
  });
});
