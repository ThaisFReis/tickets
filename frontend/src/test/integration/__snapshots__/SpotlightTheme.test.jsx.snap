// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spotlight Theme Snapshot Test > renders the application with the new Spotlight theme 1`] = `
<DocumentFragment>
  <div
    class="App bg-gray-900 text-white min-h-screen p-8"
  >
    <header
      class="App-header mb-8"
    >
      <h1
        class="text-4xl font-bold"
      >
        Concert Ticket Marketplace
      </h1>
      <nav
        class="mt-4"
      >
        <a
          class="mr-4 text-lg hover:text-blue-400"
          href="#"
        >
          Home
        </a>
        <a
          class="text-lg hover:text-blue-400"
          href="#"
          role="link"
        >
          My Tickets
        </a>
      </nav>
    </header>
    <main>
      <div>
        Loading events...
      </div>
    </main>
  </div>
</DocumentFragment>
`;
